src/typeEvaluator/optimizations.ts

```diff
@@ -62,26 +62,93 @@ export function removeDuplicateTypeNodes(typeNodes: TypeNode[]): TypeNode[] {
   return newTypeNodes
 }
 
+export function hoistDuplicateTypeNodes(src: TypeNode[]): TypeNode[] {
```

Comment:
Could we add some documentation here? What's the goal here? Is it just about de-duplicating? I'm not sure what "hoist" means in this context. This code also mutates `src` which should be highlighted. Maybe this should not even return a value, but just be a mutable function?